java/arcs/android/storage/database/DatabaseImpl.kt

```diff
@@ -108,168 +118,238 @@ class DatabaseImpl(
-                put("entity_storage_key_id", storageKeyId)
-            }
-            entity.data.forEach { (fieldName, fieldValue) ->
-                content.apply {
-                    val field = fields.getValue(fieldName)
-                    put("field_id", field.fieldId)
-                    // TODO: Handle non-primitive field values and collections.
-                    put("value_id", getPrimitiveValueId(fieldValue, field.typeId, db))
-                }
-                insertWithOnConflict(
-                    TABLE_FIELD_VALUES,
-                    null,
-                    content,
-                    SQLiteDatabase.CONFLICT_REPLACE
-                )
+    suspend fun insertOrUpdate(
+        storageKey: StorageKey,
+        entity: Entity,
+        counters: Counters? = null
+    ) = writableDatabase.useTransaction {
+        val db = this
+        // Fetch/create the entity's type ID.
+        val schemaTypeId = getSchemaTypeId(entity.schema, db, counters)
+        // Set the type ID for this storage key.
+        val storageKeyId = getEntityStorageKeyId(
+            storageKey,
+            entity.id,
+            schemaTypeId,
+            db,
+            counters
+        )
+        // Insert/update the entity's field types.
+        counters?.increment(DatabaseCounters.GET_ENTITY_FIELDS)
+        val fields = getSchemaFields(schemaTypeId, db)
+        val content = ContentValues().apply {
+            put("entity_storage_key_id", storageKeyId)
+        }
+        entity.data.forEach { (fieldName, fieldValue) ->
+            content.apply {
+                val field = fields.getValue(fieldName)
+                put("field_id", field.fieldId)
+                // TODO: Handle non-primitive field values and collections.
+                put("value_id", getPrimitiveValueId(fieldValue, field.typeId, db, counters))
             }
+            counters?.increment(DatabaseCounters.UPDATE_ENTITY_FIELD_VALUE)
+            insertWithOnConflict(
+                TABLE_FIELD_VALUES,
+                null,
+                content,
+                SQLiteDatabase.CONFLICT_REPLACE
+            )
         }
+    }
 
     @VisibleForTesting
     suspend fun insertOrUpdate(
         storageKey: StorageKey,
         data: DatabaseData.Collection,
-        dataType: DataType
+        dataType: DataType,
+        counters: Counters?
     ) = writableDatabase.useTransaction {
         val db = this
 
         // Fetch/create the entity's type ID.
-        val schemaTypeId = getSchemaTypeId(data.schema, db)
+        val schemaTypeId = getSchemaTypeId(data.schema, db, counters)
 
         // Retrieve existing collection ID for this storage key, if one exists.
+        when (dataType) {
+            DataType.Collection ->
+                counters?.increment(DatabaseCounters.GET_COLLECTION_ID)
+            DataType.Singleton ->
+                counters?.increment(DatabaseCounters.GET_SINGLETON_ID)
+            else -> Unit
```

Comment:
I think you can omit the `else` branch when the `when` keyword is used as a statement (as it is here) rather than an expression.

Reply:
Yeah, it's technically okay to omit the else, unfortunately when the `when` is used with an Enum value, the compiler likes to warn about it. And I felt it was safer - albeit a bit uglier - in this case to add the else, rather than an annotation which disabled "exhaustive when clauses" for the whole class.